### PR TITLE
BOA to USD exchange added

### PR DESCRIPTION
**Pull request by waqarRns**

### CODE REVIEW

File: tests/Boascan.test.ts [hunk 1]
```diff
@@ -931,4 +932,30 @@ describe("Test of Stoa API Server", () => {
...
+            .addSearch('amount', '1.23')
+        const response = await client.get(uri.toString());
+        let expected = { amount: 1.23, USD: 0.29427996 }
+        assert.deepStrictEqual(response.data, expected);
+    });
+
+    it("Test for convertBoaToUsd()", async () => {
+        let rate = 3450246;
+        let Boa = 0.5;
+        let exchange = new Exchange(rate)
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]:
```suggestion
        let exchange = new Exchange(rate);
```

File: tests/Boascan.test.ts [hunk 2]
```diff
@@ -931,4 +932,30 @@ describe("Test of Stoa API Server", () => {
...
+        let value = exchange.convertBoaToUsd(Boa)
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]:
```suggestion
        let value = exchange.convertBoaToUsd(Boa);
```

File: tests/Boascan.test.ts [hunk 3]
```diff
@@ -931,4 +932,30 @@ describe("Test of Stoa API Server", () => {
 
         assert.deepStrictEqual(response.data, expected);
     });
+
+    it("Test for /convert-to-usd", async () => {
+        const uri = URI(stoa_addr)
+            .directory("/convert-to-usd")
+            .addSearch('amount', '1.23')
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]:
```suggestion
            .addSearch('amount', '1.23');
```

File: package.json [hunk 4]
```diff
@@ -33,6 +33,7 @@
         "axios": "^0.21.2",
         "boa-sdk-ts": "^0.2.5",
         "boa-sodium-ts": "^0.1.1",
+        "boa-wallet-lib": "0.0.56",
```
Comment (TrustHenry) [inline package.json]:
```suggestion
```

File: src/modules/common/Exchange.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        Copyright (c) 2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+import bigDecimal from 'js-big-decimal';
+import { Amount } from 'boa-wallet-lib';
```
Comment (TrustHenry) [inline src/modules/common/Exchange.ts]: import { Amount } from "boa-sdk-ts";

File: src/modules/common/Exchange.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        return Number(usdAmount);
+    }
+
+    /**
+     * Convert Amount To USD.
+     * @param amount The amount.
+     * @returns Amount in USD.
+     */
+    public convertBalanceToUsd(amount: JSBI) {
+        return this.convertBoaToUsd(Number(Amount.toString(amount)));
```
Comment (TrustHenry) [inline src/modules/common/Exchange.ts]: https://github.com/bosagora/boa-sdk-ts/blob/ce93bd29169e98ee9a55800f6f5729eccfa9dab9/src/modules/common/Amount.ts#L298
You can use this.

File: src/modules/common/Exchange.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        let usdAmount = bigDecimal.multiply(this.exchangeRate, amount);
+        return Number(usdAmount);
+    }
+
+    /**
+     * Convert Amount To USD.
+     * @param amount The amount.
+     * @returns Amount in USD.
+     */
+    public convertBalanceToUsd(amount: JSBI) {
```
Comment (TrustHenry) [inline src/modules/common/Exchange.ts]:
```suggestion
    public convertAmountToUsd(amount: JSBI) {
```

File: src/modules/common/Exchange.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    constructor(exchangeRate: number) {
+        this.exchangeRate = exchangeRate;
+    }
+
+    /**
+     * Convert BOA to USD.
+     * @param amount The amount.
+     * @returns Amount in USD.
+     */
+    public convertBoaToUsd(amount: number): number {
```
Comment (TrustHenry) [inline src/modules/common/Exchange.ts]:
```suggestion
    public convertBoaToUsd(boa: number): number {
```

File: src/modules/common/Exchange.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public exchangeRate: number
+
+    constructor(exchangeRate: number) {
+        this.exchangeRate = exchangeRate;
+    }
+
+    /**
+     * Convert BOA to USD.
+     * @param amount The amount.
```
Comment (TrustHenry) [inline src/modules/common/Exchange.ts]:
```suggestion
     * @param boa The BOA unit amount.
```

File: tests/Boascan.test.ts [hunk 10]
```diff
@@ -745,11 +746,11 @@ describe("Test of Stoa API Server", () => {
                 total_spendable: 48799999980960,
                 total_balance: 48799999980960,
                 percentage: 0,
-                value: 0,
-                full_count: 199,
+                value: 1167549.7595444643,
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]: Please mark up to three decimal places. (floor)

File: package.json (hunk 4 above)
Comment (TrustHenry) [reply package.json]: We don't use this library in stoa.
`boa-sdk-ts`  Please use this instead.

File: src/Stoa.ts [hunk 11]
```diff
@@ -288,6 +289,7 @@ class Stoa extends WebService {
         this.app.get("/proposals/", isBlackList, this.getProposals.bind(this));
         this.app.get("/proposal/:proposal_id", isBlackList, this.getProposalById.bind(this));
         this.app.get("/validator/reward/:address", isBlackList, this.getValidatorReward.bind(this));
+        this.app.get("/convert-to-usd", isBlackList, this.convertToUSD.bind(this));
```
Comment (TrustHenry) [inline src/Stoa.ts:293]: Where is this endpoint used?

**Review by TrustHenry — CHANGES_REQUESTED** [review]

File: src/Stoa.ts (hunk 11 above)
Comment (waqarRns) [reply src/Stoa.ts:293]: this will be used for conversion of fee to $ amount.

File: tests/Boascan.test.ts [hunk 12]
```diff
@@ -693,11 +693,11 @@ describe("Test of Stoa API Server", () => {
                 total_spendable: 48799999980960,
                 total_balance: 48799999980960,
                 percentage: 0,
-                value: 0,
-                full_count: 199,
+                value: 1167549.76,
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]: Why are there two decimal places?

Comment (waqarRns) [reply tests/Boascan.test.ts]: fixed. Number(number) .Remove insignificant trailing zeros from a number

**Review by TrustHenry — APPROVED** [review]

**Pull request merged**